dev/system-test/tracing.ts

```diff
@@ -0,0 +1,656 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import * as chaiAsPromised from 'chai-as-promised';
+import {expect, use} from 'chai';
+import {describe, it, beforeEach, afterEach} from 'mocha';
+import {
+  Attributes,
+  context,
+  diag,
+  DiagConsoleLogger,
+  DiagLogLevel,
+  trace,
+  TracerProvider,
+} from '@opentelemetry/api';
+import {TraceExporter} from '@google-cloud/opentelemetry-cloud-trace-exporter';
+import {Settings} from '@google-cloud/firestore';
+import {
+  AlwaysOnSampler,
+  BatchSpanProcessor,
+  ConsoleSpanExporter,
+  InMemorySpanExporter,
+  NodeTracerProvider,
+  ReadableSpan,
+} from '@opentelemetry/sdk-trace-node';
+import {setLogFunction, Firestore} from '../src';
+import {verifyInstance} from '../test/util/helpers';
+import {
+  ATTRIBUTE_KEY_DOC_COUNT,
+  SERVICE,
+  SPAN_NAME_AGGREGATION_QUERY_GET,
+  SPAN_NAME_BATCH_COMMIT,
+  SPAN_NAME_BATCH_GET_DOCUMENTS,
+  SPAN_NAME_BULK_WRITER_COMMIT,
+  SPAN_NAME_COL_REF_ADD,
+  SPAN_NAME_COL_REF_LIST_DOCUMENTS,
+  SPAN_NAME_DOC_REF_CREATE,
+  SPAN_NAME_DOC_REF_DELETE,
+  SPAN_NAME_DOC_REF_GET,
+  SPAN_NAME_DOC_REF_LIST_COLLECTIONS,
+  SPAN_NAME_DOC_REF_SET,
+  SPAN_NAME_DOC_REF_UPDATE,
+  SPAN_NAME_PARTITION_QUERY,
+  SPAN_NAME_QUERY_GET,
+  SPAN_NAME_TRANSACTION_COMMIT,
+  SPAN_NAME_TRANSACTION_GET_AGGREGATION_QUERY,
+  SPAN_NAME_TRANSACTION_GET_DOCUMENT,
+  SPAN_NAME_TRANSACTION_GET_DOCUMENTS,
+  SPAN_NAME_TRANSACTION_GET_QUERY,
+  SPAN_NAME_TRANSACTION_RUN,
+} from '../src/telemetry/trace-util';
+import {AsyncLocalStorageContextManager} from '@opentelemetry/context-async-hooks';
+import {deepStrictEqual} from 'assert';
+
+use(chaiAsPromised);
+
+// Enable OpenTelemetry debug message for local debugging.
+diag.setLogger(new DiagConsoleLogger(), DiagLogLevel.DEBUG);
+
+// Enable Firestore debug messages for local debugging.
+setLogFunction((msg: string) => {
+  console.log(`LOG: ${msg}`);
+});
+
+// TODO(tracing): This should be moved to firestore.d.ts when we want to
+//  release the feature.
+export interface FirestoreOpenTelemetryOptions {
+  enableTracing?: boolean;
+  tracerProvider?: any;
+}
+
+interface TestConfig {
+  // In-Memory tests check trace correctness by inspecting traces in memory by
+  // utilizing InMemorySpanExporter. These tests have `e2e` set to `false`.
+  // End-to-End tests check trace correctness by querying the Google Cloud Trace
+  // backend and making sure it has recorded all the expected spans. These
+  // tests have `e2e` set to `true`.
+  e2e: boolean;
+
+  // OpenTelemetry allows registering a specific OpenTelemetry instance as the
+  // "global" instance. We should make sure that the tracing works in two
+  // circumstances: (1) the user provides us with a specific OpenTelemetry
+  // instance. (2) the user does not provide an instance and we find the
+  // global instance to use.
+  globalOpenTelemetry: boolean;
```

Comment:
`useGlobalOpenTelemetry` or `hasGlobalOpenTelemetry` since it is boolean

Reply:
Done